lookout/style/format/analyzer.py

```diff
@@ -181,34 +183,49 @@ def train(cls, ptr: ReferencePointer, config: Mapping[str, Any], data_service: D
+            files = list(sorted(files, key=lambda x: x.path))
+            if config["test_dataset_ratio"] == 0:
+                train_lines = None  # means all
+                test_lines = None  # means no lines
+            else:
+                train_lines, test_lines = FormatAnalyzer.get_train_test_split(
+                    files, config["test_dataset_ratio"],
+                    random_state=lang_config["trainable_rules"]["random_state"])
+            X_train, y_train, _ = fe.extract_features(files, lines=train_lines)
+            if test_lines:
+                X_test, y_test, _ = fe.extract_features(files, lines=train_lines)
+                _log.debug("Real test ratio is %.3f",
+                           X_test.shape[0] / (X_test.shape[0] + X_train.shape[0]))
+            X_train, selected_features = fe.select_features(X_train, y_train)
             lang_config["feature_extractor"]["selected_features"] = selected_features
             lang_config["feature_extractor"]["label_composites"] = fe.labels_to_class_sequences
             lower_bound_instances = lang_config["lower_bound_instances"]
-            if X.shape[0] < lower_bound_instances:
+            if X_train.shape[0] < lower_bound_instances:
                 _log.warning("skipped %d %s files: too few samples (%d/%d)",
-                             len(files), language, X.shape[0], lower_bound_instances)
+                             len(files), language, X_train.shape[0], lower_bound_instances)
                 continue
             _log.debug("training the rules model")
             optimizer = Optimizer(n_jobs=lang_config["n_jobs"],
                                   n_iter=lang_config["n_iter"],
                                   cv=lang_config["cv"],
                                   random_state=lang_config["trainable_rules"]["random_state"])
-            best_score, best_params = optimizer.optimize(X, y)
+            best_score, best_params = optimizer.optimize(X_train, y_train)
             _log.debug("score of the best estimator found: %.6f", best_score)
             _log.debug("params of the best estimator found: %s", str(best_params))
             _log.debug("training the model with complete data")
             lang_config["trainable_rules"].update(best_params)
             trainable_rules = TrainableRules(**lang_config["trainable_rules"],
                                              origin_config=lang_config)
-            trainable_rules.fit(X, y)
+            trainable_rules.fit(X_train, y_train)
             importances = trainable_rules.feature_importances_
             _log.debug(
                 "feature importances from %s:\n\t%s",
                 lang_config["trainable_rules"]["base_model_name"],
                 "\n\t".join("%-55s %.5E" % (fe.feature_names[i], importances[i])
                             for i in numpy.argsort(-importances)[:25] if importances[i] > 1e-5))
+            if test_lines:
```

Comment:
I think train report should always be generated.

Reply:
yeah, I agree, it should not be time-consuming, so we can do it.